### PR TITLE
Refactor documentation to make it easier for students to maintain

### PR DESCRIPTION
**Pull request by j-lum**
Move sections that students are not required to maintain to their individual files.
These sections are:
- Setting Up
- Documentation
- Testing
- Dev Ops

Include a guide on how to set up PlantUML and use it to create idiomatic diagrams.

Replace old pptx diagrams with PlantUML diagrams.

### CODE REVIEW

File: docs/DeveloperGuide.adoc [hunk 1]
```diff
@@ -272,15 +205,15 @@ If the `currentStatePointer` is at index `addressBookStateList.size() - 1`, poin
...
 
-image::UndoRedoNewCommand4StateListDiagram.png[width="800"]
+image::undo-redo_005.png[width="541px"]
 
 The following activity diagram summarizes what happens when a user executes a new command:
 
-image::UndoRedoActivityDiagram.png[width="650"]
+image::commit-activity-diagram.png[width="244px"]
 
 ==== Design Considerations
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: line 222: Alternative 1 is actually the current design (?). Note that we used a different design a few semester's ago. Most likely we forgot to update this line when we changed the design of undo/redo.

Comment (damithc) [reply docs/DeveloperGuide.adoc]: Alternative 2 says,
> `Cons: We must ensure that the implementation of each individual command are correct.`

Of course we should. :-p The real con is the implementation of each command will be more complicated.

Comment (damithc) [reply docs/DeveloperGuide.adoc]: `Aspect: Data structure to support the undo/redo commands` 
This section seems to be outdated as well.

**Comment by damithc** [comment]
Alternatively, break into separate PRs. e.g., pptx->puml can be a separate PR.

File: docs/UsingPlantUml.adoc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+***
+
+=== Exporting UML diagrams as PNGs for use
+
+As GitHub has not implemented native support for PlantUML, we export all of our diagrams as images to avoid external
+dependencies. The PlantUML integration plugin supports exporting individual files.
+
+The PlantUML executable jar provided on the official http://plantuml.com/download[PlantUML website] will export the
+entire directory instead. This can prove to be extremely useful.
```
Comment (damithc) [inline docs/UsingPlantUml.adoc]: Does this mean students have to do some manual image generation after updating the puml file?

File: docs/UsingPlantUml.adoc [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+[source, puml]
+```
+!include style.puml
+
+box Logic LOGIC_COLOR_T2
+participant ":LogicManager" as LogicManager LOGIC_COLOR
+participant ":AddressBookParser" as AddressBookParser LOGIC_COLOR
+participant ":UndoCommand" as UndoCommand LOGIC_COLOR
+end box
+```
```
Comment (damithc) [inline docs/UsingPlantUml.adoc]: Can provide code for other stylistic tweaks (to match our UML style)?

File: docs/DeveloperGuide.adoc [hunk 4]
```diff
@@ -16,82 +16,15 @@ endif::[]
 
 By: `Team SE-EDU`      Since: `Jun 2016`      Licence: `MIT`
 
-== Setting up
+== Setting Up
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: I thought the idea is to extract out this content into separate pages, not just separate files, so that when students generate the dev guide pdf file for submission, it will not contain such content. yet, the netlify preview shows that content is still in the main dev guide page https://deploy-preview-1017--seedu-ab4.netlify.com/usingplantuml

File: docs/DeveloperGuide.adoc [hunk 5]
```diff
@@ -32,7 +32,7 @@ endif::[]
 === Architecture
 
 .Architecture Diagram
-image::Architecture.png[width="600"]
+image::component-diagram.png[width="480"]
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: In the original diagram, `Main` is supposed to be connected to _everything_ (that's why the arrows are not connected to anything). If that is not reproducible in puml, you can omit the arrows altogether.

File: docs/DeveloperGuide.adoc [hunk 6]
```diff
@@ -32,7 +32,7 @@ endif::[]
...
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: Is it possible to swap location of `user` and `web` (so that the diagram can be read left to right)?

Comment (damithc) [inline docs/DeveloperGuide.adoc]: Is it possible to specify the size of the boxes, to reflect the size of each component. e.g., `Main` is a lot smaller than `Logic`

File: docs/DeveloperGuide.adoc [hunk 7]
```diff
@@ -64,23 +64,23 @@ Each of the four components
 For example, the `Logic` component (see the class diagram given below) defines it's API in the `Logic.java` interface and exposes its functionality using the `LogicManager.java` class.
 
 .Class Diagram of the Logic Component
-image::LogicClassDiagram.png[width="800"]
+image::logic-class-diagram.png[width="705"]
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: I would rather than three diagrams than show multiple levels in the same diagram. But that can be a separate PR.

File: docs/DeveloperGuide.adoc [hunk 8]
```diff
@@ -64,23 +64,23 @@ Each of the four components
...
+image::architecture-sequence-diagram.png[width="552"]
 
 The sections below give more details of each component.
 
 [[Design-Ui]]
 === UI component
 
 .Structure of the UI Component
-image::UiClassDiagram.png[width="800"]
+image::ui-class-diagram.png[width="1077"]
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: Can show incoming arrow?

File: docs/DeveloperGuide.adoc [hunk 9]
```diff
@@ -179,13 +112,13 @@ link:{repoURL}/src/main/java/seedu/address/logic/Logic.java[`Logic.java`]
 Given below is the Sequence Diagram for interactions within the `Logic` component for the `execute("delete 1")` API call.
 
 .Interactions Inside the Logic Component for the `delete 1` Command
-image::DeletePersonSdForLogic.png[width="800"]
+image::delete-sequence-diagram.png[width="1011"]
```
[the remark this answers is not in the document]
Comment (damithc) [reply docs/DeveloperGuide.adoc]: Can use even lighter colors?

File: docs/DeveloperGuide.adoc [hunk 10]
```diff
@@ -112,13 +112,13 @@ link:{repoURL}/src/main/java/seedu/address/logic/Logic.java[`Logic.java`]
...
 
 [[Design-Model]]
 === Model component
 
 .Structure of the Model Component
-image::ModelClassDiagram.png[width="800"]
+image::model-class-diagram.png[width="754px"]
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: Show incoming arrow?

File: docs/DeveloperGuide.adoc [hunk 11]
```diff
@@ -132,13 +132,13 @@ The `Model`,
...
  +
-image:ModelClassBetterOopDiagram.png[width="800"]
+image:better-model-class-diagram.png[width="598px"]
 
 [[Design-Storage]]
 === Storage component
 
 .Structure of the Storage Component
-image::StorageClassDiagram.png[width="800"]
+image::storage-class-diagram.png[width="864"]
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: Interface implementation should be shown as dashed arrow

Comment (damithc) [reply docs/DeveloperGuide.adoc]: Show incoming arrow.

File: docs/DeveloperGuide.adoc [hunk 12]
```diff
@@ -174,29 +174,29 @@ Given below is an example usage scenario and how the undo/redo mechanism behaves
 
 Step 1. The user launches the application for the first time. The `VersionedAddressBook` will be initialized with the initial address book state, and the `currentStatePointer` pointing to that single address book state.
 
-image::UndoRedoStartingStateListDiagram.png[width="800"]
+image::undo-redo.png[width="446"]
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: Add an outline for objects. Otherwise they look like text with background highlighting.

Comment (damithc) [reply docs/DeveloperGuide.adoc]: Diagrams don't seem to match original.
image
image

File: docs/DeveloperGuide.adoc [hunk 13]
```diff
@@ -205,15 +205,15 @@ If the `currentStatePointer` is at index `addressBookStateList.size() - 1`, poin
 
 Step 5. The user then decides to execute the command `list`. Commands that do not modify the address book, such as `list`, will usually not call `Model#commitAddressBook()`, `Model#undoAddressBook()` or `Model#redoAddressBook()`. Thus, the `addressBookStateList` remains unchanged.
 
-image::UndoRedoNewCommand3StateListDiagram.png[width="800"]
+image::undo-redo_004.png[width="446"]
 
 Step 6. The user executes `clear`, which calls `Model#commitAddressBook()`. Since the `currentStatePointer` is not pointing at the end of the `addressBookStateList`, all address book states after the `currentStatePointer` will be purged. We designed it this way because it no longer makes sense to redo the `add n/David ...` command. This is the behavior that most modern desktop applications follow.
 
-image::UndoRedoNewCommand4StateListDiagram.png[width="800"]
+image::undo-redo_005.png[width="541px"]
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: Try to avoid using magic numbers in file names.

File: docs/DeveloperGuide.adoc [hunk 14]
```diff
@@ -241,29 +169,29 @@ Given below is an example usage scenario and how the undo/redo mechanism behaves
...
-image::UndoRedoExecuteUndoStateListDiagram.png[width="800"]
+image::undo-redo_003.png[width="446"]
 
 [NOTE]
 If the `currentStatePointer` is at index 0, pointing to the initial address book state, then there are no previous address book states to restore. The `undo` command uses `Model#canUndoAddressBook()` to check if this is the case. If so, it will return an error to the user rather than attempting to perform the undo.
 
 The following sequence diagram shows how the undo operation works:
 
-image::UndoRedoSequenceDiagram.png[width="800"]
+image::undo-sequence-diagram.png[width="934"]
```
Comment (damithc) [inline docs/DeveloperGuide.adoc]: background boxes: use lighter colors, omit outline. Try to make them subtle as possible as they are not part of UML.

**Comment by pyokagan** [comment]
Also, I noticed that this PR changes the naming convention of the diagrams from `FooBar.png` to `foo-bar.png`.

Not saying that it absolutely can't be done, but it *is* an established convention that has been around for a very long time, so there needs to be a very good reason why it is being changed in the commit message.

**Comment by j-lum** [comment]
Been a long time coming but I've standardized the naming convention for images. I briefly considered keeping normal images as kebab-case and UML diagrams as PascalCase but the effort required to somehow pass down this convention on to students and future maintainers vs the one-off effort to standardize everything made it an easy choice. 

Spoke with @damithc and we decided on making PascalCase the standard.

The move of images into their own directory is to clean up the clutter in the images directory. Along the way a few images that were no longer used but still hanging around were identified and purged.

Snuck in a removal of the `GetStartedProgramming` section because:
- It's a lot of work to support 2 distinct set of tutorials
- Students working on AB-4 should ideally be competent enough to get by without them
- It's a lot of images

File: docs/DevOps.adoc [hunk 15]
```diff
@@ -0,0 +1,54 @@
+= AddressBook Level 4 - Developer Guide - Dev Ops
```
[the remark this answers is not in the document]
Comment (damithc) [reply docs/DevOps.adoc]: Comment pending from previous review.

File: docs/DeveloperGuide.adoc [hunk 16]
```diff
@@ -16,82 +16,15 @@ endif::[]
...
-When you are ready to start coding,
+ifdef::env-github[]
+Refer to the guide <<SettingUp#, here>>
+endif::[]
 
-1. Get some sense of the overall design by reading <<Design-Architecture>>.
-2. Take a look at <<GetStartedProgramming>>.
+ifndef::env-github[]
+include::SettingUp.adoc[tags=settingup]
+endif::[]
```
[the remark this answers is not in the document]
Comment (damithc) [reply docs/DeveloperGuide.adoc]: Comment pending from previous review.

File: docs/UsingPlantUml.adoc [hunk 17]
```diff
@@ -0,0 +1,82 @@
+= Using PlantUML
```
[the remark this answers is not in the document]
Comment (damithc) [reply docs/UsingPlantUml.adoc]: Comment pending from previous review.

File: docs/DevOps.adoc (hunk 15 above)
[the remark this answers is not in the document]
Comment (j-lum) [reply docs/DevOps.adoc]: Changed the commit message to start with an imperative present active verb.

File: docs/UsingPlantUml.adoc [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+:stylesDir: stylesheets
+:experimental:
+:toc:
+:toc-title:
+ifdef::env-github[]
+:tip-caption: :bulb:
+:note-caption: :information_source:
+endif::[]
+
+== Setting up PlantUML
```
[the remark this answers is not in the document]
Comment (j-lum) [reply docs/UsingPlantUml.adoc]: Snuck a link into the Documentation.adoc. I was actually considering having a top-level document to replace the old `GettingStartedProgramming` appendix but that's more of an AB-3.5 thing.

[the remark this answers is not in the document]
Comment (damithc) [reply docs/UsingPlantUml.adoc]: It should be in this commit. SE-EDU doesn't merge _impure_ commits. Each commit should be self-contained and does only one thing.

**Comment by j-lum** [comment]
Closing this PR in favor of the new one(s).
